### ISSUE TITLE
Steamwebhelper possibly overallocating X clients ("Maximum number of clients reachedError: Can't open display: :0")

### ISSUE THREAD

**Issue by redyoshi49q** [ValveSoftware/steam-for-linux#9561]
#### Your system information

* Steam client version Thursday, May 25th, 10:25 .p.m. UTC
* Distribution (e.g. Ubuntu): Linux Mint 20.3 Cinnamon
* Opted into Steam client beta?: Yes
* Have you checked for system updates?: Yes

#### Please describe your issue in as much detail as possible:

I'm posting this as a separate issue from #9094 rather than commenting there because this seems like it might be another lookalike bug relating to X clients (I'll explain why I think that in a moment).

I've periodically entered a state where no GUI windows would spawn.  When attempting to spawn something from the command line, an error like the following would result:

```
1478/64287 MB RAM 10.1% 9.44% 9.25% Sat, Jun 3, 2023 11:09:36 pm [1 jobs]
[ethan@firetail: +2] ~ $ xeyes &
[2] 1499175
Maximum number of clients reachedError: Can't open display: :0
[2]    1499175 exit 1     xeyes

1256/64287 MB RAM 7.56% 9.81% 10.3% Sun, Jun 4, 2023 11:54:33 pm [1 jobs]
[ethan@firetail: +2] ~ $
```

After researching the issue, I found [a helpful comment](https://bbs.archlinux.org/viewtopic.php?pid=1611548#p1611548) with a command to run in order to figure out what processes had allocated the most X clients:

```
1449/64287 MB RAM 10.6% 9.25% 9.88% Mon, Jun 5, 2023 12:10:50 am [1 jobs]
[ethan@firetail: +2] ~ $ lsof | awk '/libX11.so/ {clients[$1]++;} END {for(c in clients){printf "%s\t%s\n", clients[c], c;}}'
4       xapp-sn-w
4       csd-autom
4       csd-power
34      Web\x20Co
130     Discord
4       Utility
50      steam
36      cinnamon
120     Isolated
1       lutris-wr
27      WebExtens
3       lutris
4       mintrepor
4       csd-color
2       oosplash
19      polkit-gn
7       mintUpdat
6       csd-print
5       Socket
4       ibus-exte
6       kdeconnec
14      cinnamon-
4       gnome-ter
3       csd-clipb
4       csd-xrand
2       xprop
93      dropbox
4       csd-orien
74      itch
4       csd-xsett
4       csd-keybo
8       csd-a11y-
3       at-spi-bu
13      winedevic
4       blueman-a
117     firefox-b
20      xed
4       csd-wacom
26      Privilege
4       csd-mouse
4       nm-applet
5       csd-sound
4       ibus-ui-g
20      blueman-m
3       at-spi2-r
3       blueman-t
3       csd-scree
4       csd-backg
5       xplayer-v
32      evolution
5       csd-media
5       soffice.b
1       applet.py
4       xdg-deskt
3       redshift-
1       fontforge
67      telegram
4       redshift
4       RDD\x20Pr
21      nemo
4       nemo-desk
917     steamwebh
4       csd-house
3       csd-curso
138     palemoon
3       ibus-x11
4       goa-daemo

1375/64287 MB RAM 10.7% 9.62% 9.94% Mon, Jun 5, 2023 12:12:23 am [1 jobs]
[ethan@firetail: +2] ~
```

Based on the 917 clients associated with `steamwebh`, it seems like the most plausible culprit.  The actual `steam` process has only 50 clients in comparison (unlike what was reported in #9094, where the `steam` process was the one hoarding all of the connections).

I can produce additional log files if any might prove to be helpful.

**Comment by cstew2** [comment]
I think I'm encountering this too.
From what I can see each time I get a steam notification popup the number of unix sockets increases for Xorg and steamwebhelper and never goes down. So perhaps the windows for notifications aren't being closed correctly?

**Comment by zivkovic** [comment]
Same thing occurring to me.

Manjaro, latest stable branch.

```
 /etc/X11/xorg.conf.d  sudo ss -x src "*/tmp/.X11-unix/*" | grep -Eo "[0-9]+\s*$" | while read port                                                                                                                       ✔  9s  
do sudo ss -p -x | grep -w $port | grep -v X11-unix
done | grep -Eo '".+"' | sort | uniq -c | sort -rn
    141 "steamwebhelper"
      9 "Discord"
      8 "steam"
      5 "skypeforlinux"
      5 "chrome"
      3 "teams"
      3 "kglobalaccel5"
      3 "kded5"
      2 "python"
      2 "ksmserver"
      2 "kscreen_backend"
      2 "java"
```

**Comment by kotarou3** [comment]
Same here. steamwebhelper had almost 200 connections open with Xorg after leaving it open for a few weeks.  
Restarting steam brought it down to a more manageable ~10

**Comment by calendulish** [comment]
Same on SteamOS/SteamDeck. I have 1000 friends, for me it only takes a hour to happen

```
$ lsof -U +c 15 | cut -f1 -d' ' | sort | uniq -c | sort -rn | head -3
    297 steamwebhelper
    258 Xorg
```

**Comment by harikattar** [comment]
The lsof command is a slight overcount because it includes all sockets from steamwebhelper, not just connections to xorg.  It still shows the notification leak behavior described though.

Verified this bug is still present since I just had to restart steam to be able to use my PC.
`Steam Beta Branch:  Stable Client
Steam Version:  1716584667
Steam Client Build Date:  Fri, May 24 4:48 PM UTC -05:00
Steam Web Build Date:  Fri, May 24 4:31 PM UTC -05:00
Steam API Version:  SteamClient021`

**Comment by huntertur** [comment]
I just encountered this problem today after leaving Steam running for weeks. Restarting Steam shows client counts similar to those that @kotarou3 shared in an earlier reply.

**Comment by Lightsockie** [comment]
This happens if steam is left open for a few days for me. Luckily because of other problems with the steam client, I routinely have to `killall` steam & friends more often than that. So its a bit rare it survives that long and I run into this again lol

It seems like its a universal problem, but some more data points:
```
OS: Artix Linux x86_64
Kernel: 6.9.8-273-tkg-bore
CPU: AMD Ryzen 9 7950X
GPU: AMD ATI Radeon RX 7900 XT
WM: Awesome
```

**Comment by Novattz** [comment]
this happens regardless if i close steam or not. when running a command like:
```bash
sudo ss -x src "*/tmp/.X11-unix/*" | grep -Eo "[0-9]+\s*$" | while read port
do sudo ss -p -x | grep -w $port | grep -v X11-unix 
done | grep -Eo '".+"' | sort | uniq -c | sort -rn
```
the result showed that steamwebhelper had 224 clients open. killing steam with pkill steam -f obviously results in the issue being temporarily fixed. the problem is gonna come up again when steam has been opened one too many times because it doesnt kill the clients it creates after a new one is opened.

**Comment by Xlaits** [comment]
Adding my two cents here, just starting having this issue as well. It wasn't noticed right away as I had usually been shutting down my rig when going to bed.

**Comment by TTimo** [comment]
Hello,

We have observed this on occasion, but we have not been able to reproduce with any reliability in order to investigate.

If you are still experiencing this problem with the latest Steam beta client, please chime in and provide the usual info (https://github.com/ValveSoftware/steam-for-linux/blob/master/issue_template.md)

**Comment by suprovsky** [comment]
Steam client version: 1723685248
Distribution: Ubuntu 24.04 LTS amd64
Opted into Steam client beta: yes, Steam Families Beta
Have you checked for system updates: yes, up to date
Steam logs: [steam-logs.tar.gz](https://github.com/user-attachments/files/16647540/steam-logs.tar.gz)
GPU: Nvidia GeForce RTX 4080 SUPER

All you have to do in order to reproduce the issue on my machine is to just wait a very long time with Steam opened. After some time you will not be able to open any app including Steam games. Closing Steam and reopening it resolves the issue.

**Comment by TTimo** [comment]
Hello @suprovsky,

> All you have to do in order to reproduce the issue [..]

Do you confirm that you can still reproduce this with version `1723685248`?

**Comment by suprovsky** [comment]
No, recently it's not been happening. Let me leave my PC up and running for a couple of days and we'll see if it's a problem still.

**Comment by ahyangyi** [comment]
This bug happens for me recently, so apparently it has not gone away...

**Comment by MasterIO02** [comment]
Suddenly getting the same issue. Leaving Steam open a few hours and windows won't open anymore, the NVIDIA X server settings app becomes empty and I'm getting the error "Maximum number of clients reached" in a third-party app that regularly spawns/closes invisible windows, until I close Steam and everything goes back to normal.

If I try opening Steam again, the issue reoccurs instantly. Note that when Steam is closed, I can open as many windows as I want and the NVIDIA x server settings app is showing normally.

Linux Mint 22.3 Cinnamon

**Comment by BitBlitObviMormon** [comment]
I started getting the same issue after updating steam as well. It no longer takes a few weeks -- it only takes a few hours.
Also Linux Mint 22 Cinnamon.

**Comment by CheeseManFuu** [comment]
Also occurring on Debian 13 on linux-6.18.15 in X11. For me if the issue is already happening, closing Steam and trying to reopen it causes it to happen again instantly, making Steam and all other applications unusable until I kill all Steam processes including the Proton wineserver, but Steam and also Lutris (since it also uses Proton) won't open back up. After a computer restart it has a brief moment of stability again.

**Comment by madrang** [comment]
Maximum number of clients reached... 😢 
Got the same here after updating Steam on Ubuntu Noble.

Edit: I think it could be something else with Ubuntu, I've seen the same issue again today where no apps can create windows but Steam wasn't opened. After closing every open windows, the other apps seems to be working again...

But on the other hand, now opening Steam results in
```
[2026-05-16 18:29:17] Set status message: Verifying installation...
[----] Verifying installation...
[2026-05-16 18:29:18] Verification complete
UpdateUI: skip show logo
[2026-05-16 18:29:18] Destroy window

Steam logging initialized: directory: /home/madrang/.steam/logs

[2026-05-16 18:29:18] ProcessNextMessage: socket disconnected
[2026-05-16 18:29:18] No more messages are expected - exiting
Maximum number of clients reached
Maximum number of clients reached
Unable to open display
Unable to open displaycrash_20260516182918_5.dmp[858364]: Uploading dump (out-of-process)
/tmp/dumps/crash_20260516182918_5.dmp
crash_20260516182918_5.dmp[858364]: Finished uploading minidump (out-of-process): success = yes
crash_20260516182918_5.dmp[858364]: response: CrashID=bp-066dad42-e7e8-4c1f-b696-650af2260516
crash_20260516182918_5.dmp[858364]: file ''/tmp/dumps/crash_20260516182918_5.dmp'', upload yes: ''CrashID=bp-066dad42-e7e8-4c1f-b696-650af2260516''
/home/madrang/.steam/steam.sh: line 966: 858344 Segmentation fault      (core dumped) "$STEAMROOT/$STEAMEXEPATH" "$@"
```
When yesterday it was only spamming `Maximum number of clients reached` until I force closed it...

**Comment by superkooks** [comment]
Happening for me as well. Breaks my screenlocker too... 
Linux 7.0.5-2-MANJARO  -  X11 21.1.22-2  -  XFCE 4.20.4-1  -  steam 1.0.0.85-6

**Comment by yowshi** [comment]
also getting this error, this actually gets bad enough to take down my system such that a clean reboot is not possible and i have to use the power switch.
Linux 6.12.88+deb13-amd64 
Debian 6.12.88-1 (2026-05-15) x86_64 GNU/Linux
PRETTY_NAME="Debian GNU/Linux 13 (trixie)"
NAME="Debian GNU/Linux"
VERSION_ID="13"
VERSION="13 (trixie)"
VERSION_CODENAME=trixie
DEBIAN_VERSION_FULL=13.5
ID=debian
kwin 6.3.6
x11
Steam Beta Branch:  Stable Client
Steam Version:  1778281814
Steam Client Build Date:  Fri, May 8, 2026 18:40 UTC -08:00
Steam Web Build Date:  Fri, May 8, 2026 18:34 UTC -08:00
Steam API Version:  SteamClient023

**Comment by yowshi** [comment]
further testing revealed this is an electron application issue. mine was coming from discord, which is leaking X11 connections.

**Comment by gepandz** [comment]
> further testing revealed this is an electron application issue. mine was coming from discord, which is leaking X11 connections.

Wow, yep, killing Discord allowed Steam to start up on my system (Ubuntu 24.04 LTS, Nvidia GPU, KDE Plasma). Huh...

I was seeing the "Maximum number of clients reached" error, but not seeing "Error: Can't open display: :0" part, but this solved the problem for both Steam and the game I was trying to launch. Thanks!

**Comment by yowshi** [comment]
yeah steam does not seem to mention the display :0 part in it's console output. you just get max clients reached, mind it isnt the only application that does that. when you reach max client connections try opening something like the heroic launcher. it gives the same error in it's console output.

**Comment by CheeseManFuu** [comment]
> Wow, yep, killing Discord allowed Steam to start up on my system (Ubuntu 24.04 LTS, Nvidia GPU, KDE Plasma). Huh...
> 
> I was seeing the "Maximum number of clients reached" error, but not seeing "Error: Can't open display: :0" part, but this solved the problem for both Steam and the game I was trying to launch. Thanks!

I do want to mark the importance of the fact that by closing Discord, you opened up at least one space for an X client which allowed you to open Steam in the first place. While I don't want to discredit the idea that somebody at Discord made a fuck-up when they pushed the update that allows it to update on its own instead of requiring you to download a new `.deb` each time, it may have been coincidental and worth monitoring over time. Similarly, if it were Electron itself, I also use Element/Riot and it has not proven to be a problem.

**Comment by xpander69** [comment]
having the same issue.
restarting steam and discord solved the issue for now but it seems something is leaking over time?

to check the connections i used:

`sudo ss -x src "*/tmp/.X11-unix/*" | grep -Eo "[0-9]+\s*$" | while read port; do sudo ss -p -x | grep -w $port | grep -v X11-unix ; done | grep -Eo '".+"' | sort | uniq -c | sort -rn`

lists all the active client connections.

i wonder whats the default? 256?

edit: seems discord is the culprit thats leaking indeed.
started at 11 connections and after half an hour its 21 connections now

```
21 "DiscordCanary"
     11 "steamwebhelper"
      9 "steam"
      7 "tidal-hifi"
      3 "mate-menu"
      2 "xdg-desktop-por"
```

**Comment by yowshi** [comment]
i've been tracking usage with this command

`while true; do echo "=== $(date) ==="; sudo ss -x src "*/tmp/.X11-unix/*" | grep -Eo "[0-9]+\s*$" | while read port; do sudo ss -p -x | grep -w $port | grep -v X11-unix; done | grep -Eo '".+"' | sort | uniq -c | sort -rn | head -5; sleep 300; done | tee ~/x11-monitor.log
`
and i have been observing that discord ticks up 1 connection about every 5 minutes. this is why i am sure this isnt a steam issue.

**Comment by FlandreDaisuki** [comment]
Thanks to @yowshi 's command, I can reproduce the same result. The root cause is indeed the Discord desktop client (discord-1.0.136.deb).

<img width="1305" height="1227" alt="Image" src="https://github.com/user-attachments/assets/4941a4f9-0235-4e74-aa48-fd6f828dc14b" />

**Comment by salt-wizard** [comment]
I'm seeing the exact same behavior with @yowshi 's command. I've also tested with a .deb install and a flatpak install; Same behavior. I've left Steam running with the Discord app off and I haven't seen any issues when opening new clients.

**Comment by yowshi** [comment]
the issue is a known issue with discord here
https://github.com/flathub/com.discordapp.Discord/issues/637

**Comment by keyboardsmoke** [comment]
Great, it's a known issue. It's preventing me from logging in from sleep because Steam is using so many connections that it's preventing the login display window from coming up. Can we do something about this? It's making it such that I _cannot_ leave steam running while logged out as I have to restart my entire UI and kill steam just to get back in.

**Comment by CheeseManFuu** [comment]
> Great, it's a known issue. It's preventing me from logging in from sleep because Steam is using so many connections that it's preventing the login display window from coming up. Can we do something about this? It's making it such that I _cannot_ leave steam running while logged out as I have to restart my entire UI and kill steam just to get back in.

Did you read the above replies? Have you checked if Discord is the true issue and not Steam? I'd also check other Electron apps just in case if you aren't using the Linux Discord Desktop app.

**Comment by keyboardsmoke** [comment]
Considering steam is requesting this often to the point its flooding my logs, let's call it a little of column A and B, then. Discord still running I can restart my UI and login, whereas steam still running I can't. May be true that discord is leaking handles, but there is some other issue here on steam's side that is certainly not helping. I have no way of checking if either of these apps are at issue because the source code to neither exists, or I'd submit a patch myself. It's been pissing me off for over a week.

```
May 24 19:46:27 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:28 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:29 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:30 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:30 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:30 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:31 dt-pc steam[5958]: Maximum number of clients reached
May 24 19:46:32 dt-pc steam[5958]: Maximum number of clients reached
```

**Comment by yowshi** [comment]
> Replying to https://github.com/ValveSoftware/steam-for-linux/issues/9561#issuecomment-4530642787

yes steam has to request X11 client connections in order to launch, so do the games it launches. IF something like discord is eating all available client connections then steam cant launch and nor could any game launch from steam.
electron applications are known to have issues with leaking X11 connections, discord is probably the culprit now though it could be some other electron application.
over on the discord thread, linked in my post above, i posted a script i use to monitor and restart discord when it uses to many connections.

importantly you need to kill the application before it eats up all 256 available x11 connections, default number for linux OS's. or however many is available after all your other applications have loaded.
my script has never been tested in sleep though, you may just have to close discord completly before putting your system to sleep.

**Comment by xpander69** [comment]
the issue seems fixed on discord-canary. after running it for an hour the connection count hasnt increased yet.
i know its not a steam issue, but letting people know here. hopefully it will get into the stable discord soon also.

edit: looks like stable discord is fixed also with the JS change apparently. The actual native fix will arrive next week they say
